include/revng/Model/Types.h

```diff
@@ -0,0 +1,903 @@
+
+  static std::string key(const UpcastablePointer<T> &Val) { return Val->Name; }
+
+  static UpcastablePointer<T> fromKey(const std::string &Name) {
+    return UpcastablePointer<T>(new T(Name));
+  }
+};
+
+/// \brief A qualifier for a flatc::Type
+struct flatc::Qualifier {
+  QualifierKind::Values Kind = QualifierKind::Invalid;
+  std::size_t Size = 0;
+
+  // Kind is not Invalid, Pointer and Const have no Size, Array has Size.
+  bool verify() const debug_function {
+    switch (Kind) {
+    case QualifierKind::Invalid:
+      return false;
+    case QualifierKind::Pointer:
+    case QualifierKind::Const:
+      return not Size;
+    case QualifierKind::Array:
+      return Size;
+    }
+    return false;
+  }
+
+  static Qualifier createConst() { return { QualifierKind::Const, 0 }; }
+
+  static Qualifier createPtr() { return { QualifierKind::Pointer, 0 }; }
+
+  static Qualifier createArray(std::size_t S) {
+    return { QualifierKind::Array, S };
+  }
+
+  bool operator==(const Qualifier &) const = default;
+};
+INTROSPECTION_NS(flatc, Qualifier, Kind, Size);
+
+namespace flatc {
+
+inline bool isConstQualifier(const Qualifier &Q) {
+  revng_assert(Q.verify());
+  return Q.Kind == QualifierKind::Const;
+}
+
+inline bool isArrayQualifier(const Qualifier &Q) {
+  revng_assert(Q.verify());
+  return Q.Kind == QualifierKind::Const;
+}
+
+inline bool isPtrQualifier(const Qualifier &Q) {
+  revng_assert(Q.verify());
+  return Q.Kind == QualifierKind::Const;
+}
+
+using TypesVector = SortedVector<UpcastablePointer<flatc::Type>>;
+
+} // end namespace flatc
+
+struct flatc::TypesTree {
+  TypesVector Types;
+  bool operator==(const TypesTree &) const = default;
+  bool verify() const {
+
+    // All types on their own should verify
+    for (auto &Type : Types)
+      if (not Type->verify())
+        return false;
+
+    // FIXME: should also check that there are no loops in the type system and
+    // that there are no duplicate names
+    return true;
+  }
+};
+INTROSPECTION_NS(flatc, TypesTree, Types);
+
+namespace flatc {
+
+using Path = TupleTreeReference<flatc::Type, flatc::TypesTree>;
+using TypesTupleTree = TupleTree<TypesTree>;
+
+} // end namespace flatc
+
+/// \brief A qualified version of a flatc::Type. Can have many nested qualifiers
+struct flatc::QualifiedType {
+  Path ThePath = Path::fromString("/Types");
+  std::vector<Qualifier> Qualifiers = {};
+  // ThePath is valid, all Qualifiers are valid, and there are no consecutive
+  // Const entry in Qualifiers.
+  bool verify() const debug_function;
+  std::weak_ordering operator<=>(const QualifiedType &) const = default;
+};
+INTROSPECTION_NS(flatc, QualifiedType, ThePath, Qualifiers);
+
+namespace flatc {
+
+inline bool isQualified(const QualifiedType &QT) {
+  // FIXME: should we unwrap ThePath and look for nested qualifiers?
+  revng_assert(QT.verify());
+  return not QT.Qualifiers.empty();
+}
+
+inline constexpr std::array<const char *, 93> CReservedKeywords = {
```

Comment:
Let's push all the verification logic in the `.cpp` file.

Reply:
Done